조건문 안에서 React Hook을 호출했을 때 뜨는 ESLint 경고

## 어떤 버그인가요?

> 훅(useEffect)이 조건부로 호출되는 것으로 간주돼서 에러가 남

## 어떤 상황에서 발생한 버그인가요?

> ```C:\Users\user\MainProject\oz_externship_frontend\src\components\common\modal\BaseModal.tsx
  23:3  error  React Hook "useEffect" is called conditionally. React Hooks must be called in the exact same order in every component render  react-hooks/rules-of-hooks
  31:3  error  React Hook "useEffect" is called conditionally. React Hooks must be called in the exact same order in every component render  react-hooks/rules-of-hooks```

## 예상 결과

>  `useEffect` 훅이 조건부가 아닌 최상위에서 항상 호출되어야 함
>  모달 열림/닫힘 상태에 따라 ESC 키 이벤트와 body 스크롤 잠금이 정상 동작해야 함
>  SSR 환경에서도 안전하게 동작해야 함

--- a/src/components/common/modal/BaseModal.tsx
+++ b/src/components/common/modal/BaseModal.tsx
@@ -16,25 +16,27 @@ const BaseModal = ({
   describedById,
   panelClassName = '',
 }: BaseModalProps) => {
-  if (typeof document === 'undefined') return null
-  const root = modalUtils.getOrCreateRoot()
+  const isBrowser =
+    typeof window !== 'undefined' && typeof document !== 'undefined'
+  const root = isBrowser ? modalUtils.getOrCreateRoot() : null
 
   // ESC 닫기
   useEffect(() => {
-    if (!isOpen) return
+    if (!isBrowser || !isOpen) return
     const onKey = (e: KeyboardEvent) => e.key === 'Escape' && onClose()
     document.addEventListener('keydown', onKey)
     return () => document.removeEventListener('keydown', onKey)
-  }, [isOpen, onClose])
+  }, [isBrowser, isOpen, onClose])
 
   // body 스크롤 잠금
   useEffect(() => {
-    if (!isOpen) return
+    if (!isBrowser || !isOpen) return
     modalUtils.lockScroll()
     return () => modalUtils.unlockScroll()
-  }, [isOpen])
+  }, [isBrowser, isOpen])
 
-  if (!isOpen) return null
+  if (!isBrowser || !root || !isOpen) return null
+
   const { w, h } = MODAL_SIZES[size]
 
   return createPortal(

